### PR TITLE
Progress Report 2017 Part 1

### PR DESCRIPTION
**Pull request by anodium**
Wrote up the 2017 progress report, and added my name to the `authors.json`. Will squash all commits together after I get the banner done and LGTM'ed.

### CODE REVIEW

**Comment by jroweboy** [comment]
First off, it might be worth it to change it from "Q1" report to "since the last report" report or something. Lots happened at the end of the year and we didn't get to mention them. Some other things added that would be nice to mention:

* CPU JIT by merry and others (it was several months of work and never got a blog post. really deserves one of its own but its way too late for that)
* Framelimiter by emmaus https://github.com/citra-emu/citra/pull/2223
* Scheduler Rewrite by subv https://github.com/citra-emu/citra/pull/2260
* IPC Rewrite started by subv https://github.com/citra-emu/citra/pull/2249

I can write up piece about each of them if you want me to. I kinda want this to look like a lot has changed since a lot has!

**Comment by wwylele** [comment]
and don't forget CRO (read: pokemon) started by Subv.

**Comment by yuriks** [comment]
As is, I think it's better to leave this is a Q1 report (maybe 2016Q4 + 2017Q1) and publish it in a timely manner. I we want we can add these older topics as a "Things we missed..." articles or series of articles afterwards. There's no particular need to have articles be very long.

**Comment by chris062689** [comment]
I agree, I don't want this article to become huge. I'd rather have smaller articles we can spread out. Placing articles in small bites helps keep project awareness.

**Comment by jroweboy** [comment]
Thats fine, but some of these still happened in Q1 and were left out. Both the frame limiter and the scheduler rewrite happened in Q1 and as far as user facing changes that made games playable, these were pretty big for people.

In any case, (outside of proof reading the article which I plan to do later today) I'm fine with shipping the article as is. I'm just glad we are finally getting an article out after almost a year of no articles.

**Comment by anodium** [comment]
@jroweboy I could reword it to be part 1 of Q1, then include the PRs you mentioned in Part 2 Q1. As for 2016 Q2~4, I agree with @yuriks that we should have "What we missed" articles later on.

The CPU JIT article could still be written up, as the only time it'd be too late would be if it's obsoleted by something else, but we should focus on catching up first. :sweat_smile:

File: site/content/entry/citra-progress-report-2017-p1.md [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+
+For now, you can only configure your gamepad by manually editing a config file, but [JayFoxRox](https://github.com/jayfoxrox) has written a little tool to help edit the config file while we integrate it better into Citra's UI. You can find the tool in the Citra forums [here](https://community.citra-emu.org/t/temporary-controller-configurations-for-citra/1061).
+
+## [More Shader Refactoring Part 1](https://github.com/citra-emu/citra/pull/2335), [Part 2](https://github.com/citra-emu/citra/pull/2346), & [Part 3](https://github.com/citra-emu/citra/pull/2476) by [yuriks](https://github.com/yuriks)
+
+Citra's video code has grown organically from the start, not much was known about the PICA200 (i.e., the 3DS GPU), so a lot of the backend for it was based on pure guesswork and the team's past experience with working with emulation. It made a lot of assumptions on how it worked internally, which worked okay some of the time, but it was amazing at the time to even have a 3DS game load, let alone render a scene. So it was left alone.
```
Comment (jroweboy) [inline site/content/entry/citra-progress-report-2017-p1.md]: > It made a lot of assumptions on how it worked internally, which worked okay some of the time,

We currently don't have any pictures on the post, so how about we insert one here showing off a broken game like Animal Crossing New Leaf. Then the follow up sentence can be reworded a wee bit to make it flow better.

> It made a lot of assumptions on how it worked internally, which worked okay some of the time, but sometimes wacky things happen and its hard to figure out why. It was amazing at the time to even have a 3DS game load, let alone render a scene, but the graphics code was getting hard to maintain, so yuriks began this large refactor.

> https://cloud.githubusercontent.com/assets/272940/24393115/b2674b62-134b-11e7-898e-d5ef86e65b41.PNG (caption: Shouldn't there be a bridge here? Image courtesy of Saphiresurf)

(should probably ask saphire for permission to use the image too. I grabbed it from the issue https://github.com/citra-emu/citra/issues/2647 )

File: site/content/entry/citra-progress-report-2017-p1.md [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+
+The entire architecture of the 3DS was largely unknown, so development was done in small seperate bits at a time, and then trying to piece it all together into one cohesive piece of software. But, over time, as new pieces of information were found, the assumptions that were made previously broke. Some assumptions were very large, obvious, and easy to overcome, but a more subtle error that these assumptions created were not even in the design of the GPU, but rather the architecture and organization of the code itself.
```
Comment (jroweboy) [inline site/content/entry/citra-progress-report-2017-p1.md]: > so development was done in small seperate bits at a time, and then trying to piece it all together into one cohesive piece of software.

Suggested change: (correct spelling of separate and remove the `and` conjunction)

> so development was done in small separate bits which were pieced together into one cohesive piece of software.

File: site/content/entry/citra-progress-report-2017-p1.md [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+
+The video code was very interconnected using this interface, so despite people being aware of this, it was also extremely difficult to move code around and reorganize it, as it could very easily affect unrelated bits of code that depended on it. This was only compounded by the hardware renderer, as it was written much like earlier, slowly adding hardware support bit-by-bit to seperate parts of the video code, and simply falling back into the software renderer when disabled or not supported.
+
+There aren't many changes visible on the surface here, but internally this rewrite of the video code that [yuriks](https://github.com/yuriks) has been working on for the past few months has laid the foundation that Citra needs to be able to grow and mature in a clean and maintainable way. With these, things like geometry shaders (which many games heavily depend upon), more support for hardware renderers, and more accurate software renderers, could all be added with little to no effort besides the actual implementation.
+
+## [UDS Services](https://github.com/citra-emu/citra/pull/2628) by [Subv](https://github.com/Subv)
```
Comment (jroweboy) [inline site/content/entry/citra-progress-report-2017-p1.md]: So I talked with @Subv about this and he requested that we remove this section. He said that he wants to have it all bundled in a single, hype blog post. Instead, maybe we can stick in the frame limiter and the scheduler rewrite here (or maybe the LUT fixes which you have pictures for but nothing written about it?)

Comment (anodium) [inline site/content/entry/citra-progress-report-2017-p1.md]: Yeah, @JayFoxRox mentioned both @Subv potentially not wanting hype just yet, and I also told him that and said that the LUT images are there because I'd also written an article in 70fb0b9, but I found it to be a bit awkwardly worded, and too short, since it was a mainly internal thing.

File: site/content/entry/citra-progress-report-2017-p1.md (hunk 1 above)
Comment (anodium) [inline site/content/entry/citra-progress-report-2017-p1.md]: > We currently don't have any pictures on the post, so how about we insert one here showing off a broken game like Animal Crossing New Leaf.
Which is why I'd written up the paragraph on the LUT back on 70fb0b9.

> should probably ask saphire for permission to use the image too. I grabbed it from the issue citra-emu/citra#2647
@Saphiresurf Mind if I use this image? :stuck_out_tongue:

File: site/content/entry/citra-progress-report-2017-p1.md [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+author = "anodium"
+forum = 00
++++
+
+It's a bit hard to believe, but Citra is already [three years old](https://github.com/citra-emu/citra/commit/7a7917df27cfe5c342bba6c15b57d389fce585bd), and in these three years we've gone from an emulator that could barely run homebrew, to one that can run many commercially available games at playable speeds! It's all thanks to the hard work of volunteers working to improve the emulator at a pace we could've only imagined back in the spring of 2014, and it seems 2017 shows no signs of slowing down. We've had a lot of improvements so far with accuracy and even more 3DS features, so let's take a look.
+
+## [Gamepad Support](https://github.com/citra-emu/citra/pull/2497) by [wwylele](https://github.com/wwylele)
+
+A very often requested feature we've had on the forums and Discord server has been gamepad support, and [wwylele](https://github.com/wwylele) has finally given us just that; gamepad support, using SDL as the input backend. In order to do this, they have completely rewritten the input handling architecture, replacing it with a much cleaner interface that will allow adding different input methods much more easily in the future.
```
Comment (anodium) [inline site/content/entry/citra-progress-report-2017-p1.md]: > Perhaps the most requested feature we've had on both the forums and Discord server has been gamepad support, and wwylele has finally delivered!

I used that exact same phrasing in the draft, but changed it due to it implying that this was @wwylele's responsibility, and was slacking off.

> Using SDL as the input backend, Citra now supports a huge variety of gamepads, including XBox controllers and PS4 Controllers.

Will definitely add this, and maybe one more controller, due to the [rule of three](https://en.wikipedia.org/wiki/Rule_of_three_(writing)).

File: site/content/entry/citra-progress-report-2017-p1.md [hunk 5]
```diff
@@ -0,0 +1,33 @@
++++
+date = "2017-04-04T20:00:00-04:00"
+title = "Citra Progress Report - 2017 P1"
+tags = [ "progress-report" ]
+author = "anodium"
+forum = 00
++++
+
+It's a bit hard to believe, but Citra is already [three years old](https://github.com/citra-emu/citra/commit/7a7917df27cfe5c342bba6c15b57d389fce585bd), and in these three years we've gone from an emulator that could barely run homebrew, to one that can run many commercially available games at playable speeds! It's all thanks to the hard work of volunteers working to improve the emulator at a pace we could've only imagined back in the spring of 2014, and it seems 2017 shows no signs of slowing down. We've had a lot of improvements so far with accuracy and even more 3DS features, so let's take a look.
```
Comment (anodium) [inline site/content/entry/citra-progress-report-2017-p1.md]: Gods this sounds *so* much better, definitely adding that, since I really just put that sentence in there to essentially shove people into the meat of the article.

@JayFoxRox Also suggested to me in DMs that I invert the paragraph structure, such that "It's...speeds!" and "It's...down." are swapped. (Correct me if I interpreted your feedback wrong.)

> [6:40 PM] JayFoxRox: I think it would be better to start out with saying that 2017 is not showing signs of slowing down, many contributors [maybe also mention newbs if we had any this year?] and THEN congratulating on the 3 years
> [6:40 PM] JayFoxRox: but I'll probably try to come up with something tomorrow when I do the full review
> [6:41 PM] JayFoxRox: because right now the headline doesn't match the first sentence so my first impression was like: "wait.. is a sentence missing here? is this a full 3 year recap or just for 2017? wait... 2014? wat?"

File: site/content/entry/citra-progress-report-2017-p1.md (hunk 1 above)
Comment (Subv) [inline site/content/entry/citra-progress-report-2017-p1.md]: Good old _how did anything even work at all?_

aaah, good memories.

**Comment by Subv** [comment]
Ref the scheduler rewrite and IPC changes, they are part of a ongoing effort to allow running native system services in place of some HLE services, this can potentially help debug a lot of things in an easier way.

File: site/content/entry/citra-progress-report-2017-p1.md (hunk 1 above)
Comment (yuriks) [inline site/content/entry/citra-progress-report-2017-p1.md]: "Citra's video code has grown organically from the start, **as** not much was known about the PICA200 (i.e., the 3DS GPU) **then**, so a lot of the **emulation code** for it".

Suggestion to make the sentences flow a bit better (imo). Also replace "backend" because that might be a bit ambiguous: it could refer to the software or OpenGL  rasterizer backends, but this comment applies to the GPU code as a whole, including the shared parts.

Comment (Saphiresurf) [inline site/content/entry/citra-progress-report-2017-p1.md]: Feel free to use that photo ^_^

**Comment by jroweboy** [comment]
Not trying to sound pushy, but it would be nice to fix this up and get it live soon  💟 Q1 ended over a month ago haha

**Comment by anodium** [comment]
@jroweboy Haha, yeah sorry about that, turns out it's really easy to forget about something if you're not barraged by emails about it constantly. :sweat_smile:

**Comment by anodium** [comment]
@Subv I'll add the scheduler rewrite and IPC in a catch-up article, since the PRs are dated 2016 November and 2016 December. Fortunately they're still relevant, and I really don't want to delay this even more than it already has.

**Comment by jroweboy** [comment]
🚢 it

**Comment by bunnei** [comment]
Can we use an HD screenshot for the banner? i.e. at least 4X internal res.
The contrast ratio between the white text and the light background is also very low.

**Comment by anodium** [comment]
@bunnei I couldn't find a screenshot at that resolution that wasn't already on the Citra website, and I can't render one myself because I don't have the minimum hardware required to run Citra *at all*. If you can find one, please, go right ahead. I just want to be done with this so I can start writing the next article.

File: site/content/entry/citra-progress-report-2017-p1.md [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+
+There aren't many changes visible on the surface here, but internally this rewrite of the video code that [yuriks](https://github.com/yuriks) has been working on for the past few months has laid the foundation that Citra needs to be able to grow and mature in a clean and maintainable way. With these, things like geometry shaders (which many games heavily depend upon), more support for hardware renderers, and more accurate software renderers, could all be added with little to no effort besides the actual implementation.
+
+## Contributors of 2017
+
+Despite only being mid-April, we've already made some great progress forward this year, especially so under the hood. The sheer majority of the work already done is not user-facing by any stretch, but it will definitely make it much easier for developers to contribute additional code and makes Citra significantly easier to maintain, which of course will lead to an even faster development cycle in the future. As always, thank [you all](https://github.com/citra-emu/citra/graphs/contributors?from=2017-01-01&amp;to=2017-04-16&amp;type=c) very much for taking the time to work on Citra, and helping it become what is has, and will be.
```
Comment (Subv) [inline site/content/entry/citra-progress-report-2017-p1.md]: Not april anymore

Comment (anodium) [inline site/content/entry/citra-progress-report-2017-p1.md]: Right, lemme ninja edit that.

**Comment by anodium** [comment]
@Subv @bunnei Fixed the typo and banner, hope we can get this merged :smile:
